src/gt4py/caching.py

```diff
@@ -0,0 +1,412 @@
+
+    @property
+    def stencil_id(self) -> gt4py.definitions.StencilID:
+        """Get the id of a stencil object."""
+        fingerprint = {
+            "__main__": self.builder.definition._gtscript_["canonical_ast"],
+            "docstring": inspect.getdoc(self.builder.definition),
+            "api_annotations": f"[{', '.join(self._extract_api_annotations())}]",
+            **self._extract_externals(),
+        }
+
+        # typeignore because attrclass StencilID has generated constructor
+        return gt4py.definitions.StencilID(  # type: ignore
+            self.builder.options.qualified_name,
+            gt4py.utils.shashed_id(gt4py.utils.shashed_id(fingerprint), self.options_id),
+        )
+
+    @property
+    def module_prefix(self) -> str:
+        return "m_"
+
+    @property
+    def module_postfix(self) -> str:
+        backend_name = gt4py.utils.slugify(self.builder.backend.name)
+        id_version = self.builder.stencil_id.version
+        return f"__{backend_name}_{id_version}"
+
+    @property
+    def class_name(self) -> str:
+        name = self.builder.options.name
+        return f"{name}__{self.module_postfix}"
+
+
+class NoCachingStrategy(CachingStrategy):
+    """Apply no caching, useful for CLI."""
+
+    name = "nocaching"
+
+    def __init__(
+        self, builder: "StencilBuilder", *, output_path: pathlib.Path = pathlib.Path(".")
+    ):
+        super().__init__(builder)
+        self._output_path = output_path
+
+    @property
+    def root_path(self) -> pathlib.Path:
+        """Get stencil code output path."""
+        return self._output_path
+
+    @property
+    def backend_root_path(self) -> pathlib.Path:
+        """Simply pass through the chosen output path."""
+        return self.root_path
+
+    @property
+    def cache_info_path(self) -> Optional[pathlib.Path]:
+        """No cache info to be written."""
+        return None
+
+    def generate_cache_info(self) -> Dict[str, Any]:
+        """Nothing to generate."""
+        return {}
+
+    def update_cache_info(self) -> None:
+        """Nothing to update."""
+        pass
+
+    def validate_cache_info(self, cache_info: Dict[str, Any], *, validate_hash: bool,) -> bool:
+        """Nothing to validate, always simulate a cache miss"""
+        return False
+
+    def check_cache_info(self, *, validate_hash: bool) -> bool:
+        """Nothing to check, NoCache always validates as False."""
+        return False
+
+    @property
+    def cache_info(self) -> Dict[str, Any]:
+        """Give empty cache info."""
+        return {}
+
+    @property
+    def stencil_id(self):
+        """Get a fingerprint-less stencil id."""
+        return gt4py.definitions.StencilID(
+            qualified_name=self.builder.options.qualified_name, version=""
+        )
+
+
+def strategy_factory(name: str, builder: "StencilBuilder", *args, **kwargs) -> CachingStrategy:
+    strategies = {"jit": JITCachingStrategy, "nocaching": NoCachingStrategy}
```

Comment:
Just a comment here: if in the future, for some reason we will need additional caching strategies (or to allow the user to implement its own ones) it could be better to allow them to register new implementations instead of hardcoding them here. I don't see the need for it right now tough, so I'm fine with the current implementation.

Reply:
Agreed, this is meant to keep the barrier to introducing a registering mechanism low without over-engineering for now.